subsys/pm/pm_cpu_shell.c

```diff
@@ -0,0 +1,173 @@
+
+	const struct pm_state_info *state_info;
+
+	/* sanity check for first */
+	if (residency_info_count == 0) {
+		return -EINVAL;
+	}
+
+	shell_print(sh, "Supported Low Power States:");
+
+	for (state_info = residency_info;
+			state_info < (residency_info + ARRAY_SIZE(residency_info)); state_info++) {
+		shell_print(sh,
+		"  - State: %s, Substate: %d, Residency: %dus, Latency: %dus, PM Device Disabled: %s",
+			pm_state_to_str(state_info->state),
+			state_info->substate_id,
+			state_info->min_residency_us,
+			state_info->exit_latency_us,
+			state_info->pm_device_disabled ? "Yes" : "No");
+	}
+
+	return 0;
+}
+
+static int cmd_cpu_available(const struct shell *sh, size_t argc, char **argv)
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+
+	const struct pm_state_info *state_info;
+	bool available;
+	bool locked;
+
+	shell_print(sh, "Check whether the low power states of the current core are supported:");
+
+	for (state_info = residency_info;
+			state_info < (residency_info + ARRAY_SIZE(residency_info)); state_info++) {
+		available = pm_policy_state_is_available(state_info->state,
+				state_info->substate_id);
+		locked = pm_policy_state_lock_is_active(state_info->state,
+				state_info->substate_id);
+
+		shell_print(sh, " - %-16s sub=%-3u avail=%c lock=%c",
+				pm_state_to_str(state_info->state),
+				state_info->substate_id,
+				available ? 'Y' : 'N',
+				locked ? 'Y' : 'N');
+	}
+
+	return 0;
+}
+
+static int cmd_cpu_idle(const struct shell *sh, size_t argc, char **argv)
+{
+	if (argc < 4) {
+		shell_error(sh, "Usage: cpu idle times(ms) <state> [substate (0..255)]");
+		return -EINVAL;
+	}
+
+	int err;
+	enum pm_state st;
+	uint8_t sub;
+	const struct pm_state_info *cpu_state = NULL;
+	uint32_t ms;
+
+	ms = shell_strtoul(argv[1], 0, &err);
+	if (err < 0) {
+		shell_error(sh, "Unable to parse input (err %d), times", err);
+		return err;
+	}
+
+	if (pm_state_from_str(argv[2], &st)) {
+		shell_error(sh, "Unknown state: %s", argv[2]);
+		return -EINVAL;
+	}
+
+	sub = shell_strtoul(argv[3], 0, &err);
+	if (err < 0) {
+		shell_error(sh, "Unable to parse input (err %d), substate", err);
+		return err;
+	}
+
+	/* sanity check firstly if sleep time is more than min-residency */
+	for (size_t i = 0; i < residency_info_count; i++) {
+		if ((residency_info[i].state == st) &&
+				(residency_info[i].substate_id == sub)) {
+			cpu_state = &residency_info[i];
+			break;
+		}
+	}
+
+	if (cpu_state == NULL) {
+		shell_error(sh,
+				"State %s substate %u is not supported on this CPU",
+				argv[2], sub);
+		return -ENOTSUP;
+	}
+
+
+	/* check if time is available for min residency */
+	if (ms * 1000 < cpu_state->min_residency_us) {
+		shell_error(sh,
+				"Sleep %u ms (%u us) < min-residency %u us for %s:%u",
+				ms, ms * 1000,
+				cpu_state->min_residency_us,
+				pm_state_to_str(cpu_state->state),
+				cpu_state->substate_id);
+		return -EINVAL;
+	}
+
+	pm_policy_state_lock_put(st, sub);
+
+	k_msleep(ms);
+
+	pm_policy_state_lock_get(st, sub);
```

Comment:
This can't guarantee entering the state/substate set, unless every other sleep state has been locked, and nothing else is holding this state/substate lock.

The shell really only needs the ability to lock/unlock sleep states, and list which states are and are not locked. Once the state locks are configured, the system will idle, the k_msleep() here should be redundant